File: src/scan/probe.ts
import net from 'node:net';
import http from 'node:http';
import https from 'node:https';
import type { IncomingMessage } from 'node:http';
import type { TLSSocket } from 'node:tls';
import type { HttpInfo, PageMeta, Protocol, TlsInfo } from '../shared/types.ts';
import { detectFramework, extractMeta } from './metadata.ts';

/** Only the head matters, and unbounded reads on a log-streaming endpoint hurt. */
const MAX_BODY_BYTES = 64 * 1024;
// Tuned against real silent listeners (Apple's rapportd accepts and then says
// nothing). Anything that has not spoken by now is not going to.
const SNIFF_TIMEOUT = 800;
// Generous for loopback, where a healthy server answers in single-digit ms.
const REQUEST_TIMEOUT = 2500;
const MAX_REDIRECTS = 3;
/**
 * Ceiling on a single port's entire probe, redirects included. Without it a
 * chain of slow hops can hold a probe worker for far longer than any one
 * request timeout suggests.
 *
 * Every genuine HTTP server measured on a real machine replied inside 600ms,
 * so this leaves ample headroom. The ports that reach it are ones like a
 * non-HTTP service that accepts the socket and then stalls - they are going to
 * end up classified `tcp` regardless, so waiting longer buys nothing.
 */
const PROBE_DEADLINE = 3000;

export interface ProbeResult {
  protocol: Protocol;
  http?: HttpInfo;
  meta?: PageMeta;
  tls?: TlsInfo;
  probeMs: number;
  error?: string;
}

/**
 * Decide what a port is speaking by sending a plaintext HTTP request and
 * looking at the first bytes of the reply.
 *
 *   "HTTP/"      -> plaintext HTTP
 *   0x16 / 0x15  -> TLS handshake or alert; our cleartext GET was not a valid
 *                   ClientHello, so a TLS server answers with an alert
 *   RST, no data -> many TLS stacks just drop garbage; worth a TLS retry
 *   anything else-> some other protocol, or a banner server like SSH/SMTP
 *
 * Returns a hint rather than a verdict; `probePort` acts on it.
 */
function sniff(
  port: number,
  host: string,
  signal?: AbortSignal,
): Promise<'http' | 'maybe-tls' | 'other' | 'closed'> {
  return new Promise((resolve) => {
    if (signal?.aborted) return resolve('closed');

    const socket = new net.Socket();
    let settled = false;
    let received = Buffer.alloc(0);
    let connected = false;

    const finish = (result: 'http' | 'maybe-tls' | 'other' | 'closed') => {
      if (settled) return;
      settled = true;
      signal?.removeEventListener('abort', onAbort);
      socket.removeAllListeners();
      socket.destroy();
      resolve(result);
    };

    const classify = () => {
      if (received.length === 0) {
        // Connected, accepted our request, said nothing. Either a client-speaks-
        // first protocol or a TLS server that ignored us. Worth one TLS attempt.
        return finish(connected ? 'maybe-tls' : 'closed');
      }
      const first = received[0]!;
      if (first === 0x16 || first === 0x15) return finish('maybe-tls');

      const head = received.subarray(0, 5).toString('latin1');
      if (head.startsWith('HTTP/')) return finish('http');

      return finish('other');
    };

    const onAbort = () => finish('closed');
    signal?.addEventListener('abort', onAbort, { once: true });

    socket.setTimeout(SNIFF_TIMEOUT);

    socket.once('connect', () => {
      connected = true;
      socket.write(`GET / HTTP/1.1\r\nHost: ${host}:${port}\r\nConnection: close\r\n\r\n`);
    });

    socket.on('data', (chunk: Buffer) => {
      received = Buffer.concat([received, chunk]);
      // The verdict only needs the first few bytes.
      if (received.length >= 8) classify();
    });

    socket.once('timeout', () => (connected ? classify() : finish('closed')));
    socket.once('end', classify);
    socket.once('close', classify);
    socket.once('error', (err: NodeJS.ErrnoException) => {
      if (!connected) return finish('closed');
      // A reset after a successful connect is the classic "you spoke the wrong
      // protocol at me" signal.
      if (err.code === 'ECONNRESET') return finish('maybe-tls');
      classify();
    });

    socket.connect(port, host);
  });
}

interface FetchOutcome {
  status: number;
  statusText: string;
  headers: Record<string, string>;
  body: string;
  tls?: TlsInfo;
  finalUrl: string;
  redirectTo?: string;
  firstStatus: number;
}

/** GET a URL, reading at most MAX_BODY_BYTES and following a few redirects. */
async function fetchHead(
  url: string,
  secure: boolean,
  signal?: AbortSignal,
  depth = 0,
  firstStatus?: number,
  firstRedirect?: string,
): Promise<FetchOutcome> {
  const outcome = await new Promise<FetchOutcome>((resolve, reject) => {
    const mod = secure ? https : http;
    const req = mod.request(
      url,
      {
        method: 'GET',
        // Dev servers overwhelmingly use self-signed or mkcert certs; refusing
        // them would make the tool useless for exactly the case it exists for.
        ...(secure ? { rejectUnauthorized: false } : {}),
        headers: {
          // Some servers content-negotiate; ask for HTML so we get a <title>.
          accept: 'text/html,application/xhtml+xml,*/*;q=0.8',
          'user-agent': 'ports-scanner/0.1 (+localhost inventory)',
          connection: 'close',
        },
        timeout: REQUEST_TIMEOUT,
        signal,
      },
      (res: IncomingMessage) => {
        const chunks: Buffer[] = [];
        let size = 0;

        res.on('data', (chunk: Buffer) => {
          size += chunk.length;
          if (size <= MAX_BODY_BYTES) {
            chunks.push(chunk);
          } else {
            // We have all the head we are going to need; stop paying for the rest.
            chunks.push(chunk.subarray(0, chunk.length - (size - MAX_BODY_BYTES)));
            res.destroy();
          }
        });

        const done = () => {
          const headers: Record<string, string> = {};
          for (const [k, v] of Object.entries(res.headers)) {
            if (v === undefined) continue;
            headers[k.toLowerCase()] = Array.isArray(v) ? v.join(', ') : String(v);
          }

          let tls: TlsInfo | undefined;
          if (secure) tls = readCert(res.socket as TLSSocket);

          resolve({
            status: res.statusCode ?? 0,
            statusText: res.statusMessage ?? '',
            headers,
            body: Buffer.concat(chunks).toString('utf8'),
            ...(tls ? { tls } : {}),
            finalUrl: url,
            firstStatus: firstStatus ?? res.statusCode ?? 0,
            ...(firstRedirect ? { redirectTo: firstRedirect } : {}),
          });
        };

        res.on('end', done);
        res.on('close', done);
        res.on('error', reject);
      },
    );

    req.on('timeout', () => req.destroy(new Error('timeout')));
    req.on('error', reject);
    req.end();
  });

  // Follow redirects so the index shows the app's real title rather than the
  // empty body of a 302 to /login.
  const location = outcome.headers['location'];
  if (
    location &&
    outcome.status >= 300 &&
    outcome.status < 400 &&
    depth < MAX_REDIRECTS
  ) {
    let next: URL;
    try {
      next = new URL(location, url);
    } catch {
      return outcome;
    }
    // Only follow within localhost; an external redirect is a finding in itself.
    if (isLoopback(next.hostname)) {
      try {
        return await fetchHead(
          next.toString(),
          next.protocol === 'https:',
          signal,
          depth + 1,
          outcome.firstStatus,
          firstRedirect ?? location,
        );
      } catch {
        return outcome;
      }
    }
  }

  return outcome;
}

function isLoopback(hostname: string): boolean {
  return (
    hostname === 'localhost' ||
    hostname === '127.0.0.1' ||
    hostname === '::1' ||
    hostname === '[::1]' ||
    hostname.endsWith('.localhost')
  );
}

/** Certificate DN fields are string | string[] when a name repeats. */
function firstOf(value: string | string[] | undefined): string | undefined {
  if (Array.isArray(value)) return value[0];
  return value;
}

function readCert(socket: TLSSocket): TlsInfo | undefined {
  try {
    const cert = socket.getPeerCertificate?.();
    if (!cert || Object.keys(cert).length === 0) return undefined;

    const subject = firstOf(cert.subject?.CN) ?? firstOf(cert.subject?.O);
    const issuer = firstOf(cert.issuer?.CN) ?? firstOf(cert.issuer?.O);
    const altNames = cert.subjectaltname
      ?.split(',')
      .map((s) => s.trim().replace(/^DNS:|^IP Address:/, ''))
      .filter(Boolean);

    return {
      ...(subject ? { subject } : {}),
      ...(issuer ? { issuer } : {}),
      ...(cert.valid_from ? { validFrom: cert.valid_from } : {}),
      ...(cert.valid_to ? { validTo: cert.valid_to } : {}),
      // Same CN on both ends, or no issuer at all, means nobody vouched for it.
      selfSigned: !issuer || issuer === subject,
      ...(socket.getProtocol?.() ? { protocol: socket.getProtocol()! } : {}),
      ...(altNames?.length ? { altNames } : {}),
    };
  } catch {
    return undefined;
  }
}

/** Full probe of a single open port: classify it, then describe it. */
export async function probePort(
  port: number,
  host = '127.0.0.1',
  outerSignal?: AbortSignal,
): Promise<ProbeResult> {
  const started = Date.now();
  const hostForUrl = host.includes(':') ? `[${host}]` : host;

  // Combine the caller's cancellation with our own deadline, so a pathological
  // port cannot pin a probe worker indefinitely.
  const deadline = AbortSignal.timeout(PROBE_DEADLINE);
  const signal = outerSignal
    ? AbortSignal.any([outerSignal, deadline])
    : deadline;

  const hint = await sniff(port, host, signal);

  if (hint === 'closed') {
    return { protocol: 'tcp', probeMs: Date.now() - started, error: 'no response' };
  }

  // No plaintext fallback on the maybe-tls path: the sniff already sent a
  // cleartext GET and it did not produce an HTTP reply, so repeating it can
  // only ever burn another full request timeout before failing the same way.
  const attempts: Array<{ secure: boolean }> =
    hint === 'maybe-tls'
      ? [{ secure: true }]
      : // 'http', or 'other' where the sniff saw a banner - one cheap
        // plaintext attempt, then we call it tcp.
        [{ secure: false }];

  let lastError: string | undefined;

  for (const { secure } of attempts) {
    const url = `${secure ? 'https' : 'http'}://${hostForUrl}:${port}/`;
    try {
      const res = await fetchHead(url, secure, signal);

      const httpInfo: HttpInfo = {
        status: res.firstStatus || res.status,
        ...(res.statusText ? { statusText: res.statusText } : {}),
        ...(res.redirectTo ? { redirectTo: res.redirectTo } : {}),
        // Recorded only when we actually followed the chain, which fetchHead
        // does only within loopback.
        ...(res.status !== res.firstStatus
          ? { finalStatus: res.status, finalUrl: res.finalUrl }
          : {}),
        ...(res.headers['server'] ? { server: res.headers['server'] } : {}),
        ...(res.headers['x-powered-by'] ? { poweredBy: res.headers['x-powered-by'] } : {}),
        ...(res.headers['content-type'] ? { contentType: res.headers['content-type'] } : {}),
      };
      const framework = detectFramework(res.headers);
      if (framework) httpInfo.framework = framework;

      const isHtml = (res.headers['content-type'] ?? '').includes('html') ||
        /<html|<!doctype html/i.test(res.body.slice(0, 512));
      const meta = isHtml ? extractMeta(res.body, res.finalUrl) : undefined;

      return {
        protocol: secure ? 'https' : 'http',
        http: httpInfo,
        ...(meta ? { meta } : {}),
        ...(res.tls ? { tls: res.tls } : {}),
        probeMs: Date.now() - started,
      };
    } catch (err) {
      lastError = err instanceof Error ? err.message : String(err);
    }
  }

  return {
    protocol: 'tcp',
    probeMs: Date.now() - started,
    ...(lastError ? { error: lastError } : {}),
  };
}

File: src/scan/probe.rs
//! Work out what a port is actually speaking, then describe it.
//!
//! Everything here is bounded. A scanner that hangs on one pathological
//! listener is worse than one that gives up on it, so every step has a budget
//! and the whole probe has a deadline over the top.

use std::collections::HashMap;
use std::net::IpAddr;
use std::sync::Arc;
use std::time::{Duration, Instant};

use http_body_util::BodyExt;
use hyper::{Request, Uri};
use hyper_util::rt::TokioIo;
use tokio::io::{AsyncReadExt, AsyncWriteExt};
use tokio::net::TcpStream;
use tokio_rustls::rustls::client::danger::{
    HandshakeSignatureValid, ServerCertVerified, ServerCertVerifier,
};
use tokio_rustls::rustls::pki_types::{CertificateDer, ServerName, UnixTime};
use tokio_rustls::rustls::{ClientConfig, DigitallySignedStruct, SignatureScheme};
use tokio_rustls::TlsConnector;

use crate::scan::metadata::{detect_framework, extract_meta};
use crate::types::{HttpInfo, PageMeta, Protocol, TlsInfo};

/// Only the head matters, and unbounded reads on a log-streaming endpoint hurt.
const MAX_BODY_BYTES: usize = 64 * 1024;
/// Tuned against real silent listeners (Apple's rapportd accepts and then says
/// nothing). Anything that has not spoken by now is not going to.
const SNIFF_TIMEOUT: Duration = Duration::from_millis(800);
/// Generous for loopback, where a healthy server answers in single-digit ms.
const REQUEST_TIMEOUT: Duration = Duration::from_millis(2500);
const MAX_REDIRECTS: usize = 3;
/// Ceiling on a single port's entire probe, redirects included.
///
/// Every genuine HTTP server measured on a real machine replied inside 600ms,
/// so this leaves ample headroom. The ports that reach it are ones like a
/// non-HTTP service that accepts the socket and then stalls — they end up
/// classified `tcp` regardless, so waiting longer buys nothing.
const PROBE_DEADLINE: Duration = Duration::from_millis(3000);

#[derive(Debug, Default)]
pub struct ProbeResult {
    pub protocol: Protocol,
    pub http: Option<HttpInfo>,
    pub meta: Option<PageMeta>,
    pub tls: Option<TlsInfo>,
    pub probe_ms: u64,
    pub error: Option<String>,
}

/// What the first bytes off the socket suggest we are talking to.
#[derive(Debug, PartialEq, Eq)]
enum Hint {
    Http,
    MaybeTls,
    Other,
    Closed,
}

/// Decide what a port is speaking by sending a plaintext request and looking at
/// the first bytes of the reply.
///
///   "HTTP/"      -> plaintext HTTP
///   0x16 / 0x15  -> TLS handshake or alert; our cleartext GET was not a valid
///                   ClientHello, so a TLS server answers with an alert
///   reset, no data -> many TLS stacks just drop garbage; worth a TLS retry
///   anything else  -> some other protocol, or a banner server like SSH/SMTP
async fn sniff(port: u16, host: IpAddr) -> Hint {
    let Ok(Ok(mut socket)) =
        tokio::time::timeout(SNIFF_TIMEOUT, TcpStream::connect((host, port))).await
    else {
        return Hint::Closed;
    };

    let request = format!("GET / HTTP/1.1\r\nHost: {host}:{port}\r\nConnection: close\r\n\r\n");
    if socket.write_all(request.as_bytes()).await.is_err() {
        // Connected and then refused our bytes: a TLS server is the likeliest
        // explanation, so it is worth one handshake attempt.
        return Hint::MaybeTls;
    }

    let mut buffer = [0u8; 16];
    match tokio::time::timeout(SNIFF_TIMEOUT, socket.read(&mut buffer)).await {
        Ok(Ok(0)) | Err(_) => Hint::MaybeTls,
        Ok(Err(err)) if err.kind() == std::io::ErrorKind::ConnectionReset => {
            // The classic "you spoke the wrong protocol at me" signal.
            Hint::MaybeTls
        }
        Ok(Err(_)) => Hint::MaybeTls,
        Ok(Ok(n)) => {
            let head = &buffer[..n];
            if head[0] == 0x16 || head[0] == 0x15 {
                Hint::MaybeTls
            } else if head.starts_with(b"HTTP/") {
                Hint::Http
            } else {
                Hint::Other
            }
        }
    }
}

/// Accepts any certificate.
///
/// Dev servers overwhelmingly use self-signed or mkcert certs; refusing them
/// would make this tool useless for exactly the case it exists for. We are
/// reading a certificate to describe it, not to trust anything it says.
#[derive(Debug)]
struct AcceptAnyCert;

impl ServerCertVerifier for AcceptAnyCert {
    fn verify_server_cert(
        &self,
        _end_entity: &CertificateDer<'_>,
        _intermediates: &[CertificateDer<'_>],
        _server_name: &ServerName<'_>,
        _ocsp_response: &[u8],
        _now: UnixTime,
    ) -> Result<ServerCertVerified, tokio_rustls::rustls::Error> {
        Ok(ServerCertVerified::assertion())
    }

    fn verify_tls12_signature(
        &self,
        _message: &[u8],
        _cert: &CertificateDer<'_>,
        _dss: &DigitallySignedStruct,
    ) -> Result<HandshakeSignatureValid, tokio_rustls::rustls::Error> {
        Ok(HandshakeSignatureValid::assertion())
    }

    fn verify_tls13_signature(
        &self,
        _message: &[u8],
        _cert: &CertificateDer<'_>,
        _dss: &DigitallySignedStruct,
    ) -> Result<HandshakeSignatureValid, tokio_rustls::rustls::Error> {
        Ok(HandshakeSignatureValid::assertion())
    }

    fn supported_verify_schemes(&self) -> Vec<SignatureScheme> {
        vec![
            SignatureScheme::RSA_PKCS1_SHA256,
            SignatureScheme::RSA_PKCS1_SHA384,
            SignatureScheme::RSA_PKCS1_SHA512,
            SignatureScheme::ECDSA_NISTP256_SHA256,
            SignatureScheme::ECDSA_NISTP384_SHA384,
            SignatureScheme::ECDSA_NISTP521_SHA512,
            SignatureScheme::RSA_PSS_SHA256,
            SignatureScheme::RSA_PSS_SHA384,
            SignatureScheme::RSA_PSS_SHA512,
            SignatureScheme::ED25519,
        ]
    }
}

fn tls_connector() -> TlsConnector {
    let config = ClientConfig::builder()
        .dangerous()
        .with_custom_certificate_verifier(Arc::new(AcceptAnyCert))
        .with_no_client_auth();
    TlsConnector::from(Arc::new(config))
}

/// Read what the peer's certificate claims about itself.
fn describe_cert(certs: &[CertificateDer<'_>], protocol: Option<String>) -> Option<TlsInfo> {
    use x509_parser::prelude::*;

    let leaf = certs.first()?;
    let (_, cert) = X509Certificate::from_der(leaf.as_ref()).ok()?;

    let common_name = |name: &X509Name| -> Option<String> {
        name.iter_common_name()
            .next()
            .and_then(|cn| cn.as_str().ok())
            .map(str::to_string)
            .or_else(|| {
                name.iter_organization()
                    .next()
                    .and_then(|o| o.as_str().ok())
                    .map(str::to_string)
            })
    };

    let subject = common_name(cert.subject());
    let issuer = common_name(cert.issuer());

    let alt_names: Vec<String> = cert
        .subject_alternative_name()
        .ok()
        .flatten()
        .map(|ext| {
            ext.value
                .general_names
                .iter()
                .filter_map(|name| match name {
                    GeneralName::DNSName(dns) => Some(dns.to_string()),
                    GeneralName::IPAddress(bytes) => match bytes.len() {
                        4 => Some(
                            std::net::Ipv4Addr::new(bytes[0], bytes[1], bytes[2], bytes[3])
                                .to_string(),
                        ),
                        _ => None,
                    },
                    _ => None,
                })
                .collect()
        })
        .unwrap_or_default();

    Some(TlsInfo {
        // Nobody vouched for it if the issuer matches the subject, or there is
        // no issuer at all.
        self_signed: issuer.is_none() || issuer == subject,
        subject,
        issuer,
        valid_from: Some(cert.validity().not_before.to_string()),
        valid_to: Some(cert.validity().not_after.to_string()),
        protocol,
        alt_names: if alt_names.is_empty() {
            None
        } else {
            Some(alt_names)
        },
    })
}

struct FetchOutcome {
    status: u16,
    headers: HashMap<String, String>,
    body: String,
    tls: Option<TlsInfo>,
    final_url: String,
}

/// One request/response exchange, reading at most `MAX_BODY_BYTES` of body.
async fn fetch_once(url: &Uri, secure: bool) -> anyhow::Result<FetchOutcome> {
    let host = url.host().ok_or_else(|| anyhow::anyhow!("no host"))?;
    let port = url.port_u16().unwrap_or(if secure { 443 } else { 80 });

    // A bracketed IPv6 literal in the URI needs unwrapping before it can be
    // handed to the resolver.
    let connect_host = host.trim_start_matches('[').trim_end_matches(']');
    let stream = TcpStream::connect((connect_host, port)).await?;

    // Origin-form: just the path. Handing hyper the absolute URI makes it emit
    // proxy-form (`GET http://host/ HTTP/1.1`), which is legal but routes
    // differently on servers that match the raw target — Chrome's DevTools
    // endpoint answers 404 to it, so we would misreport a healthy service.
    let target = url.path_and_query().map(|pq| pq.as_str()).unwrap_or("/");

    let request = Request::builder()
        .method("GET")
        .uri(target)
        .header("host", format!("{host}:{port}"))
        // Some servers content-negotiate; ask for HTML so we get a <title>.
        .header("accept", "text/html,application/xhtml+xml,*/*;q=0.8")
        .header("user-agent", "ports-scanner/0.2 (+localhost inventory)")
        .header("connection", "close")
        .body(String::new())?;

    let (response, tls) = if secure {
        let server_name = match connect_host.parse::<IpAddr>() {
            Ok(ip) => ServerName::IpAddress(ip.into()),
            Err(_) => ServerName::try_from(connect_host.to_string())?,
        };
        let stream = tls_connector().connect(server_name, stream).await?;

        let (_, connection) = stream.get_ref();
        let tls = describe_cert(
            connection.peer_certificates().unwrap_or(&[]),
            connection.protocol_version().map(|v| format!("{v:?}")),
        );

        let (mut sender, conn) =
            hyper::client::conn::http1::handshake(TokioIo::new(stream)).await?;
        tokio::spawn(async move {
            let _ = conn.await;
        });
        (sender.send_request(request).await?, tls)
    } else {
        let (mut sender, conn) =
            hyper::client::conn::http1::handshake(TokioIo::new(stream)).await?;
        tokio::spawn(async move {
            let _ = conn.await;
        });
        (sender.send_request(request).await?, None)
    };

    let status = response.status().as_u16();
    let headers = response
        .headers()
        .iter()
        .map(|(name, value)| {
            (
                name.as_str().to_lowercase(),
                value.to_str().unwrap_or_default().to_string(),
            )
        })
        .collect();

    // Stop reading once we have all the head we could need; a dev server
    // streaming logs from `/` would otherwise never finish.
    let mut body = response.into_body();
    let mut collected: Vec<u8> = Vec::new();
    while let Some(frame) = body.frame().await {
        let Ok(frame) = frame else { break };
        if let Some(chunk) = frame.data_ref() {
            let remaining = MAX_BODY_BYTES.saturating_sub(collected.len());
            if remaining == 0 {
                break;
            }
            let take = chunk.len().min(remaining);
            collected.extend_from_slice(&chunk[..take]);
            if collected.len() >= MAX_BODY_BYTES {
                break;
            }
        }
    }

    Ok(FetchOutcome {
        status,
        headers,
        body: String::from_utf8_lossy(&collected).into_owned(),
        tls,
        final_url: url.to_string(),
    })
}

fn is_loopback_host(host: &str) -> bool {
    let host = host.trim_start_matches('[').trim_end_matches(']');
    host == "localhost"
        || host == "127.0.0.1"
        || host == "::1"
        || host.ends_with(".localhost")
        || host
            .parse::<IpAddr>()
            .map(|ip| ip.is_loopback())
            .unwrap_or(false)
}

/// GET a URL, following a few redirects so long as they stay on this machine.
///
/// An external redirect is left unfollowed and recorded as such: a localhost
/// port that bounces you to a public site is a finding, not something to chase.
async fn fetch_head(
    start: Uri,
    secure: bool,
) -> anyhow::Result<(FetchOutcome, u16, Option<String>)> {
    let mut url = start;
    let mut secure = secure;
    let mut first_status: Option<u16> = None;
    let mut first_redirect: Option<String> = None;

    for _ in 0..=MAX_REDIRECTS {
        let outcome = fetch_once(&url, secure).await?;
        let status = outcome.status;
        first_status.get_or_insert(status);

        let location = outcome.headers.get("location").cloned();
        let Some(location) = location.filter(|_| (300..400).contains(&status)) else {
            return Ok((outcome, first_status.unwrap_or(status), first_redirect));
        };

        // Resolve against the current URL so a relative Location works.
        let Ok(base) = url::Url::parse(&url.to_string()) else {
            return Ok((outcome, first_status.unwrap_or(status), first_redirect));
        };
        let Ok(next) = base.join(&location) else {
            return Ok((outcome, first_status.unwrap_or(status), first_redirect));
        };
        if !next.host_str().map(is_loopback_host).unwrap_or(false) {
            return Ok((outcome, first_status.unwrap_or(status), first_redirect));
        }

        first_redirect.get_or_insert(location);
        secure = next.scheme() == "https";
        let Ok(parsed) = next.as_str().parse::<Uri>() else {
            return Ok((outcome, first_status.unwrap_or(status), first_redirect));
        };
        url = parsed;
    }

    // Ran out of redirect budget; re-fetch is not worth it, report what we have.
    let outcome = fetch_once(&url, secure).await?;
    let status = first_status.unwrap_or(outcome.status);
    Ok((outcome, status, first_redirect))
}

/// Full probe of a single open port: classify it, then describe it.
pub async fn probe_port(port: u16, host: IpAddr) -> ProbeResult {
    let started = Instant::now();
    let elapsed = |start: Instant| start.elapsed().as_millis() as u64;

    match tokio::time::timeout(PROBE_DEADLINE, probe_inner(port, host)).await {
        Ok(mut result) => {
            result.probe_ms = elapsed(started);
            result
        }
        Err(_) => ProbeResult {
            protocol: Protocol::Tcp,
            probe_ms: elapsed(started),
            error: Some("probe deadline exceeded".into()),
            ..Default::default()
        },
    }
}

async fn probe_inner(port: u16, host: IpAddr) -> ProbeResult {
    let hint = sniff(port, host).await;

    if hint == Hint::Closed {
        return ProbeResult {
            protocol: Protocol::Tcp,
            error: Some("no response".into()),
            ..Default::default()
        };
    }

    // No plaintext fallback on the TLS path: the sniff already sent a cleartext
    // GET and it did not produce an HTTP reply, so repeating it can only burn
    // another request timeout before failing the same way.
    let secure = hint == Hint::MaybeTls;
    let scheme = if secure { "https" } else { "http" };
    let host_for_url = if host.is_ipv6() {
        format!("[{host}]")
    } else {
        host.to_string()
    };

    let Ok(uri) = format!("{scheme}://{host_for_url}:{port}/").parse::<Uri>() else {
        return ProbeResult {
            protocol: Protocol::Tcp,
            error: Some("could not build url".into()),
            ..Default::default()
        };
    };

    match tokio::time::timeout(REQUEST_TIMEOUT, fetch_head(uri, secure)).await {
        Ok(Ok((outcome, first_status, redirect_to))) => {
            let mut http = HttpInfo {
                status: if first_status == 0 {
                    outcome.status
                } else {
                    first_status
                },
                redirect_to,
                server: outcome.headers.get("server").cloned(),
                powered_by: outcome.headers.get("x-powered-by").cloned(),
                content_type: outcome.headers.get("content-type").cloned(),
                framework: detect_framework(&outcome.headers),
                ..Default::default()
            };

            // Only recorded when we actually followed the chain, which
            // `fetch_head` does only within loopback.
            if outcome.status != http.status {
                http.final_status = Some(outcome.status);
                http.final_url = Some(outcome.final_url.clone());
            }

            let is_html = outcome
                .headers
                .get("content-type")
                .map(|ct| ct.contains("html"))
                .unwrap_or(false)
                || {
                    let head = &outcome.body[..outcome.body.len().min(512)].to_lowercase();
                    head.contains("<html") || head.contains("<!doctype html")
                };

            let meta = is_html.then(|| extract_meta(&outcome.body, &outcome.final_url));

            ProbeResult {
                protocol: if secure {
                    Protocol::Https
                } else {
                    Protocol::Http
                },
                http: Some(http),
                meta,
                tls: outcome.tls,
                ..Default::default()
            }
        }
        Ok(Err(err)) => ProbeResult {
            protocol: Protocol::Tcp,
            error: Some(err.to_string()),
            ..Default::default()
        },
        Err(_) => ProbeResult {
            protocol: Protocol::Tcp,
            error: Some("timeout".into()),
            ..Default::default()
        },
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use tokio::io::AsyncWriteExt;
    use tokio::net::TcpListener;

    /// A server that answers every connection with a fixed payload.
    ///
    /// It must keep accepting: a probe makes two connections to a port, one to
    /// sniff the protocol and a second to actually fetch.
    async fn serve(payload: &'static str) -> u16 {
        let listener = TcpListener::bind("127.0.0.1:0").await.unwrap();
        let port = listener.local_addr().unwrap().port();
        tokio::spawn(async move {
            while let Ok((mut socket, _)) = listener.accept().await {
                tokio::spawn(async move {
                    let mut scratch = [0u8; 1024];
                    let _ = socket.read(&mut scratch).await;
                    let _ = socket.write_all(payload.as_bytes()).await;
                    let _ = socket.shutdown().await;
                });
            }
        });
        port
    }

    #[tokio::test]
    async fn describes_a_plain_http_server() {
        let body = "<html><head><title>Acme</title></head><body>hi</body></html>";
        let payload: &'static str = Box::leak(
            format!(
                "HTTP/1.1 200 OK\r\ncontent-type: text/html\r\nserver: nginx\r\n\
                 content-length: {}\r\nconnection: close\r\n\r\n{body}",
                body.len()
            )
            .into_boxed_str(),
        );

        let port = serve(payload).await;
        let result = probe_port(port, IpAddr::from([127, 0, 0, 1])).await;

        assert_eq!(result.protocol, Protocol::Http);
        let http = result.http.expect("http info");
        assert_eq!(http.status, 200);
        assert_eq!(http.server.as_deref(), Some("nginx"));
        assert_eq!(http.framework.as_deref(), Some("nginx"));
        assert_eq!(result.meta.and_then(|m| m.title).as_deref(), Some("Acme"));
    }

    /// Capture the raw bytes of the second request a probe makes.
    ///
    /// The first connection is the sniff, which sends a hand-written request;
    /// the second is the real fetch, which is the one worth pinning down.
    #[tokio::test]
    async fn sends_an_origin_form_target_and_exactly_one_host_header() {
        let listener = TcpListener::bind("127.0.0.1:0").await.unwrap();
        let port = listener.local_addr().unwrap().port();
        let captured = Arc::new(std::sync::Mutex::new(Vec::<String>::new()));
        let sink = Arc::clone(&captured);

        tokio::spawn(async move {
            while let Ok((mut socket, _)) = listener.accept().await {
                let sink = Arc::clone(&sink);
                tokio::spawn(async move {
                    let mut buffer = vec![0u8; 4096];
                    if let Ok(n) = socket.read(&mut buffer).await {
                        sink.lock()
                            .unwrap()
                            .push(String::from_utf8_lossy(&buffer[..n]).into_owned());
                    }
                    let _ = socket
                        .write_all(
                            b"HTTP/1.1 200 OK\r\ncontent-length: 0\r\nconnection: close\r\n\r\n",
                        )
                        .await;
                    let _ = socket.shutdown().await;
                });
            }
        });

        probe_port(port, IpAddr::from([127, 0, 0, 1])).await;

        let requests = captured.lock().unwrap();
        let fetch = requests.last().expect("a fetch request was made");

        // Origin-form. Absolute-form is for proxies, and servers that route on
        // the raw target — Chrome's DevTools endpoint among them — 404 on it.
        assert!(
            fetch.starts_with("GET / HTTP/1.1"),
            "expected origin-form request line, got: {:?}",
            fetch.lines().next()
        );

        let host_headers = fetch
            .lines()
            .filter(|line| line.to_lowercase().starts_with("host:"))
            .count();
        assert_eq!(host_headers, 1, "duplicate Host header in:\n{fetch}");
    }

    #[tokio::test]
    async fn classifies_a_silent_listener_as_tcp() {
        // Accepts the connection, then says nothing — Apple's rapportd, and a
        // good few IPC endpoints, behave exactly like this.
        let listener = TcpListener::bind("127.0.0.1:0").await.unwrap();
        let port = listener.local_addr().unwrap().port();
        tokio::spawn(async move {
            if let Ok((socket, _)) = listener.accept().await {
                tokio::time::sleep(Duration::from_secs(30)).await;
                drop(socket);
            }
        });

        let result = probe_port(port, IpAddr::from([127, 0, 0, 1])).await;
        assert_eq!(result.protocol, Protocol::Tcp);
    }

    #[tokio::test]
    async fn a_closed_port_is_tcp_with_no_response() {
        let listener = TcpListener::bind("127.0.0.1:0").await.unwrap();
        let port = listener.local_addr().unwrap().port();
        drop(listener);

        let result = probe_port(port, IpAddr::from([127, 0, 0, 1])).await;
        assert_eq!(result.protocol, Protocol::Tcp);
        assert_eq!(result.error.as_deref(), Some("no response"));
    }

    #[test]
    fn loopback_detection_covers_the_forms_a_redirect_can_use() {
        assert!(is_loopback_host("localhost"));
        assert!(is_loopback_host("127.0.0.1"));
        assert!(is_loopback_host("[::1]"));
        assert!(is_loopback_host("myapp.localhost"));
        assert!(!is_loopback_host("example.com"));
        assert!(!is_loopback_host("10.0.0.1"));
    }
}
